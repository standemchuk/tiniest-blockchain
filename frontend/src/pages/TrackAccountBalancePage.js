import React, { Component } from 'react'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import Card from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import CardContent from '@material-ui/core/CardContent'
import { withStyles } from '@material-ui/core/styles'

import ErrorModal from '../components/ErrorModal'

import GenericTable from '../components/GenericTable'
import PageHeader from '../components/PageHeader'

const styles = theme => ({
  card: {
    margin: theme.spacing.unit * 2
  }
})

export class TrackAccountBalancePage extends Component {
  state = {
    userId: '',
    accountBalance: 0,
    userTransactions: [],
    requestError: false
  }

  async componentDidMount () {
    await this.getBalance()
  }

  getBalance = async () => {
    const userId = localStorage.getItem('userId')

    try {
      const response = await fetch(`http://localhost:3000/api/balance/${userId}`)

      if (response && response.status === 200) {
        const responseJson = await response.json()

        this.setState({
          userId,
          accountBalance: responseJson.accountBalance,
          userTransactions: responseJson.userTransactions
        })
      } else {
        this.setState({ requestError: true })
      }
    } catch (err) {
      this.setState({ requestError: true })
    }
  }

  handleModalClose = () => {
    this.setState({ requestError: false })
  }

  handleRetry = async () => {
    await this.getBalance()
  }

  render () {
    const { classes } = this.props

    return (
      <Grid container>
        <PageHeader title='Your account Balance' />
        <Grid item xs={12} sm={5}>
          <Card className={classes.card}>
            <CardHeader title='Account info' />
            <CardContent>
              <Typography paragraph>Your user id is: {this.state.userId}</Typography>
              <Typography>You have {this.state.accountBalance} coins on your account</Typography>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} sm={7}>
          <Card className={classes.card}>
            <CardHeader title='Transactions list' />
            <CardContent>
              <GenericTable
                data={this.state.userTransactions}
                headers={['Sender', 'Receiver', 'Amount', 'Created at']}
              />
            </CardContent>
          </Card>
        </Grid>
        <ErrorModal
          open={this.state.requestError}
          handleRetry={this.handleRetry}
          handleModalClose={this.handleModalClose}
        />
      </Grid>
    )
  }
}

export default withStyles(styles)(TrackAccountBalancePage)
