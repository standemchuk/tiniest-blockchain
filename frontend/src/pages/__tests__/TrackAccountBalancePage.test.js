import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import { TrackAccountBalancePage } from '../TrackAccountBalancePage'

beforeEach(() => {
  fetch.resetMocks()
})

it('renders correctly', () => {
  const tree = renderer
    .create(
      <TrackAccountBalancePage classes={{ card: 'card' }} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should fetch the data and set requestError to false if successful', () => {
  const tree = shallow(
    <TrackAccountBalancePage classes={{ card: 'card' }} />
  )

  expect(fetch).toBeCalled()
  expect(tree.state().requestError).toBe(false)
})
