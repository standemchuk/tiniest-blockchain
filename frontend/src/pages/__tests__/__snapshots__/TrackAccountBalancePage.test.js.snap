// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiGrid-container-1"
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-40"
  >
    <p
      className="MuiTypography-root-98 MuiTypography-display1-102 MuiTypography-paragraph-116 MuiTypography-alignCenter-111 Component-header-97"
    >
      Your account Balance
    </p>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-sm-5-47"
  >
    <div
      className="MuiPaper-root-123 MuiPaper-elevation1-126 MuiPaper-rounded-124 MuiCard-root-122 card"
    >
      <div
        className="MuiCardHeader-root-150"
      >
        <div
          className="MuiCardHeader-content-153"
        >
          <span
            className="MuiTypography-root-98 MuiTypography-headline-103 MuiCardHeader-title-154"
          >
            Account info
          </span>
        </div>
      </div>
      <div
        className="MuiCardContent-root-156"
      >
        <p
          className="MuiTypography-root-98 MuiTypography-body1-107 MuiTypography-paragraph-116"
        >
          Your user id is: 
          
        </p>
        <p
          className="MuiTypography-root-98 MuiTypography-body1-107"
        >
          You have 
          0
           coins on your account
        </p>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-12-40 MuiGrid-grid-sm-7-49"
  >
    <div
      className="MuiPaper-root-123 MuiPaper-elevation1-126 MuiPaper-rounded-124 MuiCard-root-122 card"
    >
      <div
        className="MuiCardHeader-root-150"
      >
        <div
          className="MuiCardHeader-content-153"
        >
          <span
            className="MuiTypography-root-98 MuiTypography-headline-103 MuiCardHeader-title-154"
          >
            Transactions list
          </span>
        </div>
      </div>
      <div
        className="MuiCardContent-root-156"
      >
        <div
          className="MuiGrid-container-1"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-40"
          >
            <h3
              className="MuiTypography-root-98 MuiTypography-subheading-105"
              style={
                Object {
                  "marginLeft": 30,
                }
              }
            >
              No data
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
